### ISSUE TITLE
sudo ratbagd outputs wierd error

### ISSUE THREAD

**Issue by mrcush** [libratbag/libratbag#535]
First off, this ~~was~~ **is** the most painful installation I've ever done.  I cloned and ran the build, from there I had to work at every error I encountered 1 by 1 using google or trial and error.  **So I'll explain my steps:**

`git clone https://github.com/libratbag/libratbag.git`
`cd libratbag`
`meson builddir --prefix=/usr/`

**This is where I started encountering errors so:**

`sudo apt-get install gcc`
`sudo apt-get install clang`
`sudo apt-get install -y pkg-config`
`sudo apt-get install ninja-build`
`sudo apt-get install libudev-dev`
`sudo apt-get install libevdev-dev`
`sudo apt-get install libglib2.0-dev`
`sudo apt-get install libsystemd-dev`
`sudo apt-get install check-devel`
`sudo apt-get install valgrind`
`sudo apt-get install swig`
`sudo apt-get install python3-dev`
`sudo apt-get install python3-pip`
`sudo apt-get install python-dev python-pip gcc`
`sudo pip3 install evdev`
`sudo pip3 install --user evdev`
`sudo pip3 install --user meson`

**YEEEESSS!!  build finished with no errors.  The relief was unimaginable.  Anyways.  The next steps:**

`sudo ninja -C builddir`
`sudo ninja -C builddir install`
`sudo ninja install`

**Tells me no ninja.build in directory so I look in builddir and see one there so:**

`cd builddir`
`sudo ninja install`

**These next 3 commands do nothing visibly in the terminal.  I enter the commands and it immediately goes back to $ for input.**

`sudo systemctl daemon-reload`
`sudo systemctl reload dbus.service`
`sudo systemctl enable ratbagd.service`

**I then started installing piper**

`git clone https://github.com/libratbag/piper.git`
`cd piper`
`sudo meson builddir --prefix=/usr/`

**ran into error so:**

`sudo apt install python-gi-dev`

**build again**

`sudo meson builddir --prefix=/usr/`

**yay, works, so continue:**

`sudo ninja -C builddir`
`sudo ninja -C builddir install`

**SO, I run:**

`sudo ratbagd`

**and get this error:**

`ratbagd error: Failed to start ratbagd: File exists`

**I have no idea what to do.  This error seems odd to me because it says, "File exists".   If I were to disect this error I would have thought it would say, "File does not exist" so that I would have to find and download it.  I have found one other person via google who has had this error but he/she received 2 downvotes and no response so their post did nothing for me but let me know that at least one other person has received the same error.  That person's post was here: https://github.com/libratbag/piper/issues/66#issuecomment-352289875**

 **I can't figure it out.  I've also tried running:**

`sudo ratbagctl`

**and get this output:**

`event17:   Logitech Gaming Mouse G502`

If I try to run piper (that has an image of a mouse trap which I've read is bad), an instance of it runs for around 10 seconds and closes and I get an error report.
If anyone can help out I would greatly appreciate it.  The steps I have shown are cut down to the major events of progress.  I have spent probably around 7 hours on this so far and my progress has been riddled with TONS of google searches, terminal commands that didn't work, the sudo ratbagd command outputting: "event17:   Logitech Gaming Mouse G502" and stalling so that I had to ctrl+C out of it until I finally found some other solutions and eventually landed where inputting sudo ratbagd actually gave me an output without stalling.  I would really love to get this working as the dpi on my mouse Logitech G502 is frustratingly low and changing it in windows and coming back to Ubuntu does not seem to hold to the dpi I set it at.  Mouse acceleration is also an issue and I would love to be able to edit buttons and so forth without having to jump through hoops.   Editing in xinput doesn't work and was told to try this.  

Thanks!
Matt

**Comment by FFY00** [comment]
You had a prebuilt binary available.

https://packages.debian.org/search?searchon=names&keywords=libratbag

Even if you didn't want to use that you could quickly check the dependencies and install everything, this way the build would run flawlessly.

Regarding to your error.
Well, you enabled the service but didn't actually start it.

```
sudo systemctl start ratbagd.service
```

Seems a little strange you can start ratbagd directly. Maybe because of DBus?

**Comment by Hjdskes** [comment]
> Seems a little strange you can start ratbagd directly. Maybe because of DBus?

This isn't strange at all, `ratbagd` is simply an executable in `/usr/bin` or `/usr/local/bin`.

>Regarding to your error. Well, you enabled the service but didn't actually start it.

This isn't the error. Ratbagd returns that error when it is already running. @DaCush must have accidentally started ratbagd somehow before running `sudo ratbagd` (probably by trying Piper first). This is also shown in the fact that `ratbagctl` successfully shows the mouse.

@DaCush You shouldn't have to start `ratbagd` manually. Simply enable the service (which you did) and it should start on request, for example when you start Piper.

> If I try to run piper (that has an image of a mouse trap which I've read is bad), an instance of it runs for around 10 seconds and closes and I get an error report.

Can you show the error report?

**Comment by FFY00** [comment]
> This isn't strange at all, `ratbagd` is simply an executable in `/usr/bin` or `/usr/local/bin`.
I meant *can't* :sweat_smile:. And by start it wasn't just to run it, but to properly start it.

**Comment by Hjdskes** [comment]
That makes more sense :)

**Comment by whot** [comment]
> First off, this was is the most painful installation I've ever done

I'd appreciate any hints on how to make this more usable but note that you're not installing an application, you're **building it from source**. If you just want to install something, use the package manager tools of your distribution.

**Comment by DaCush** [comment]
> You had a prebuilt binary available.

I wish I knew that was there.  I'm relatively new to linux and it's terminal atm.  Just got hired on for a C++ programming job and they are running linux.  I don't start until the 26th but figured I'd get a head start on learning it and getting comfortable with it so I don't look like a complete fool on the first day of the job.  Can see myself switching over completely if I enjoy it enough.  I'm also running Ubuntu and not Debian.  Hopefully that's not a requirement?  ha

> Regarding to your error.
Well, you enabled the service but didn't actually start it.

Actually, i have run that command.  I must have missed it while going through my terminal to see what to post.  Running it now give no output in case it should.  Trying to run piper after running the command does nothing as well.

> Can you show the error report?

I was going to in my post but of course when I needed to post it I couldn't get the error to pop up again.  I'll edit this here in a moment if I can get it to show.  Currently, I click Piper and nothing happens.  

>  If you just want to install something, use the package manager tools of your distribution.

Gotcha, that most likely has to do with me being new to linux.  I'll get there eventually.  Apologize if I sounded harsh in my post, I had been going at it for 7 hours originally thinking it was like any other application I've installed so far.

**Comment by whot** [comment]
run `sudo ratbagd --verbose` and check the output (after killing any running instances). If it doesn't list your device, it's not detected by ratbagd. In which case piper won't be able to handle it either, it's just a GUI frontend for ratbagd after all. Piper should also display the actual error message under the mousetrap issue, explaining what the issue is.

**Comment by DaCush** [comment]
ratbag debug: Logitech Gaming Mouse G502 is device '/dev/hidraw0'.
ratbag debug: Logitech Gaming Mouse G502 is device '/dev/hidraw1'.
ratbag debug: report ID 01
ratbag debug: report ID 03
ratbag debug: report ID 04
ratbag debug: report ID 10
ratbag debug: report ID 11
ratbag debug: 'Logitech Gaming Mouse G502' is using protocol v4.2
ratbag debug: device has color effects
> run sudo ratbagd --verbose and check the output (after killing any running instances). If it doesn't list your device, it's not detected by ratbagd. In which case piper won't be able to handle it either, it's just a GUI frontend for ratbagd after all. Piper should also display the actual error message under the mousetrap issue, explaining what the issue is.

ratbag debug: device has non-rgb leds
ratbag debug: device has adjustable dpi
ratbag debug: device is at 1200 dpi (variable between 50 and 12000).
ratbag debug: device has adjustable report rate
ratbag debug: device has 4 report rates
ratbag debug: device has onboard profiles
ratbag debug: device has configurable mouse button spy
ratbag debug: driver match found: Logitech HID++2.0
event6: "Logitech Gaming Mouse G502", 3 profiles

**Comment by whot** [comment]
in that case and after you start ratbagd, start piper and it should show the device. if not, that's a piper bug.

**Comment by DaCush** [comment]
>in that case and after you start ratbagd, start piper and it should show the device. if not, that's a piper bug.

In that case, it's a piper bug.  However, when I do start ratbagd, it hangs in the terminal.  Not sure if that's normal or not.  Can't use terminal again until I ^C

**Comment by whot** [comment]
ratbagd is quiet by default and doesn't output anything, so that behaviour is expected.

**Comment by Hjdskes** [comment]
Does Piper print any error messages when you start it from the terminal with `piper`?

**Comment by FFY00** [comment]
> I wish I knew that was there. I'm relatively new to linux and it's terminal atm. Just got hired on for a C++ programming job and they are running linux. I don't start until the 26th but figured I'd get a head start on learning it and getting comfortable with it so I don't look like a complete fool on the first day of the job. Can see myself switching over completely if I enjoy it enough. I'm also running Ubuntu and not Debian. Hopefully that's not a requirement? ha

No problem :smile:. If you just started using linux, looks like you're pretty good already. Ubuntu is actually debian based, they use the same package manager. Debian packages **should** work under ubuntu, in most cases, but if you're new to this I would recommend to not go there just yet.

You also have a package available under Ubuntu.
https://packages.ubuntu.com/search?keywords=libratbag&searchon=names&suite=bionic&section=all

Just a tip, you can easily search for packages in the CLI using the following command.
```
apt-cache search libratbag
```
(note, no sudo required)

If you have any issues please feel free to contact me, I would be happy to help :grin:.

**Comment by DaCush** [comment]
>Does Piper print any error messages when you start it from the terminal with piper?

YES!  It does.  This is the output:

`Traceback (most recent call last):
  File "/usr/bin/piper", line 36, in <module>
    from piper.application import Application
  File "/usr/lib/python3/dist-packages/piper/application.py", line 18, in <module>
    from .window import Window
  File "/usr/lib/python3/dist-packages/piper/window.py", line 21, in <module>
    from .mouseperspective import MousePerspective
  File "/usr/lib/python3/dist-packages/piper/mouseperspective.py", line 19, in <module>
    from .buttonspage import ButtonsPage
  File "/usr/lib/python3/dist-packages/piper/buttonspage.py", line 20, in <module>
    from .mousemap import MouseMap
  File "/usr/lib/python3/dist-packages/piper/mousemap.py", line 20, in <module>
    from lxml import etree
ModuleNotFoundError: No module named 'lxml'`

Seems I need a module named lxml?  I'll look into getting that installed and report back here in a moment.

>No problem smile. If you just started using linux, looks like you're pretty good already. Ubuntu is actually debian based, they use the same package manager. Debian packages should work under ubuntu, in most cases, but if you're new to this I would recommend to not go there just yet.

You also have a package available under Ubuntu.
https://packages.ubuntu.com/search?keywords=libratbag&searchon=names&suite=bionic&section=all

Just a tip, you can easily search for packages in the CLI using the following command.

apt-cache search libratbag

(note, no sudo required)

If you have any issues please feel free to contact me, I would be happy to help grin.
(email is in my profile)

Thanks!  I really appreciate that.

**Comment by FFY00** [comment]
It seems you are missing `python-lxml`.

> Thanks! I really appreciate that. Before I installed my first distro a couple days ago, I had gone to one CTF meeting where we installed Putty and I learned a small bit of Linux commands.

Cool!

**Comment by DaCush** [comment]
>It seems you are missing python-lxml.

I tired:

sudo apt-get install python-lxml`

Started piper and got the same error message.  So I looked on google for someone with same error and they were told to install:

`apt-get install python-dev libxml2 libxml2-dev libxslt-dev`

Did that, yet again same error.  I'm going to try restarting?  So I'll be right back with an update.

**Comment by DaCush** [comment]
New post because looking at the Piper crash report details, it gives something specific.  It says:

>piper crashed with modulenotfounderror in /usr/lib/python3/dist-packages/piper/mousemap.py: no module named lxml

SO, maybe installing lmxl through python is in the wrong python directory?  Maybe I need to install it specifically for python3?  I'll give it a shot

**Comment by FFY00** [comment]
Oh, yes. Forgot debian still treats `python` as `python2`. The correct package should be `python3-lxml`.

**Comment by DaCush** [comment]
Everything looks to be working.  Only thing I can't seem to figure out is how to assign buttons on my shift button, piper calls it "second mode", that I can hold for a whole new config while holding the button down.  I would assume there's a way if piper has a name for it?

Last thing:  the terminal does give an error when piper shuts down.  It shuts down the moment I press apply to apply my button changes.  I'm assuming that it is either intentional as it's a timeout error or it's a bug.  This is the output from the terminal once I hit apply in piper and it immediately shuts down:

`Timeout was reached`
`Traceback (most recent call last):`
 ` File "/usr/lib/python3/dist-packages/piper/ratbagd.py", line 193, in callback_handler`
 `   res = obj.call_finish(result)`
`GLib.Error: g-io-error-quark: Timeout was reached (24)`

As far as this issue topic, it can be marked as solved unless you'd like to keep it open for the last error I received.  Thanks a bunch for the responses everyone.

**Comment by bentiss** [comment]
Just a quick comment here. After re-reading the thread, we are clearly at fault as we haven't switched the README from a developer's perspective to a user one. When the project was not known, it was fine, but we should probably not burden new users to a full compilation when there is an easy solution.

**Comment by whot** [comment]
Closing. The original error message was fixed in a3de0e1ff79b45fd87fd79cf570258cdfe966af1 and there's a PR for #538 and there's #563 for the error in https://github.com/libratbag/libratbag/issues/535#issuecomment-420125202.
